internal/blocksync/synchronizer.go

```diff
@@ -0,0 +1,460 @@
+func NewSynchronizer(start int64, client BlockClient, blockExec *blockApplier, opts ...OptionFunc) *Synchronizer {
+	peerStore := NewInMemPeerStore()
+	logger := log.NewNopLogger()
+	bp := &Synchronizer{
+		logger:         logger,
+		clock:          clockwork.NewRealClock(),
+		client:         client,
+		applier:        blockExec,
+		peerStore:      peerStore,
+		jobGen:         newJobGenerator(start, logger, client, peerStore),
+		startHeight:    start,
+		height:         start,
+		workerPool:     workerpool.New(poolWorkerSize, workerpool.WithLogger(logger)),
+		pendingToApply: map[int64]BlockResponse{},
+	}
+	for _, opt := range opts {
+		opt(bp)
+	}
+	bp.BaseService = *service.NewBaseService(logger, "Synchronizer", bp)
+	return bp
+}
+
+// OnStart implements service.Service by spawning requesters routine and recording
+// synchronizer's start time.
+func (s *Synchronizer) OnStart(ctx context.Context) error {
+	s.lastAdvance = s.clock.Now()
+	s.lastHundredBlock = s.lastAdvance
+	s.workerPool.Run(ctx)
+	go s.runHandler(ctx, s.produceJob)
+	go s.runHandler(ctx, s.consumeJobResult)
+	return nil
+}
+
+func (s *Synchronizer) OnStop() {
+	s.workerPool.Stop(context.Background())
+}
+
+func (s *Synchronizer) produceJob(ctx context.Context) {
+	if !s.jobGen.shouldJobBeGenerated() {
+		// TODO should we stop producer loop ?
+		// TODO need to come up with a smarter way how to produce jobs without sleeping
+		s.clock.Sleep(50 * time.Millisecond)
+		return
+	}
+	// remove timed out peers and redo its heights again
+	s.removeTimedoutPeers(ctx)
+	s.jobProgressCounter.Add(1)
+	job, err := s.jobGen.nextJob(ctx)
+	if err != nil {
+		s.logger.Error("cannot create a next job", "error", err)
+		return
+	}
+	s.peerStore.PeerUpdate(job.peer.peerID, ResetMonitor(), AddNumPending(1))
+	err = s.workerPool.Send(ctx, job)
+	if err != nil {
+		s.logger.Error("cannot add a job to worker-pool", "error", err)
+	}
+}
+
+func (s *Synchronizer) consumeJobResult(ctx context.Context) {
+	res, err := s.workerPool.Receive(ctx)
+	if err != nil {
+		s.logger.Error("cannot receive a job result from worker pool", "error", err)
+		return
+	}
+	s.jobProgressCounter.Add(-1)
+	if res.Err != nil {
+		var bfErr *errBlockFetch
+		if !errors.As(res.Err, &bfErr) {
+			return
+		}
+		s.jobGen.pushBack(bfErr.height)
+		s.RemovePeer(bfErr.peerID)
+		_ = s.client.Send(ctx, p2p.PeerError{NodeID: bfErr.peerID, Err: bfErr})
+		return
+	}
+	resp := res.Value.(*BlockResponse)
+	s.peerStore.PeerUpdate(resp.PeerID, AddNumPending(-1), UpdateMonitor(resp.Block.Size()))
+	err = s.addBlock(*resp)
+	if err != nil {
+		s.logger.Error("cannot add a block to the pending list",
+			"height", resp.Block.Height,
+			"error", err.Error())
+		_ = s.client.Send(ctx, p2p.PeerError{NodeID: resp.PeerID, Err: err})
+		return
+	}
+	err = s.applyBlock(ctx)
+	if err != nil {
+		s.logger.Error("cannot apply a block", "height", resp.Block.Height, "error", err.Error())
+		s.RemovePeer(resp.PeerID)
```

Comment:
only if all peers are malicious or broken
we keep only healthy peers to request blocks
in terms of peer score, - I don't know how it can be useful here